File: app/src/main/java/com/vyy/sekerimremake/features/chart/presenter/master/ChartAdapter.java
package com.vyy.sekerimremake.features.chart.presenter.master;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.view.ViewCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.vyy.sekerimremake.R;
import com.vyy.sekerimremake.databinding.ChartDayBinding;
import com.vyy.sekerimremake.features.chart.domain.model.ChartDayModel;
import com.vyy.sekerimremake.features.chart.utils.GlucoseLevelChecker;

import java.util.List;
import java.util.Objects;

public class ChartAdapter extends RecyclerView.Adapter<ChartAdapter.ChartViewHolder> {

    private final Context mContext;
    private final List<ChartDayModel> chartDays;
    private final OnDayClickListener mOnDayListener;

    public ChartAdapter(Context mContext, List<ChartDayModel> chartDays, OnDayClickListener onDayClickListener) {
        this.mContext = mContext;
        this.chartDays = chartDays;
        this.mOnDayListener = onDayClickListener;
    }

    public class ChartViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ChartDayBinding binding;
        OnDayClickListener onDayClickListener;

        public ChartViewHolder(@NonNull ChartDayBinding b, OnDayClickListener onDayClickListener) {
            super(b.getRoot());
            binding = b;
            this.onDayClickListener = onDayClickListener;
            binding.chartDay.setOnClickListener(this);
        }

        public void setTransitionName(String imageId) {
            ViewCompat.setTransitionName(binding.chartDay, imageId);
        }

        @Override
        public void onClick(View v) {
            onDayClickListener.onRowClick(chartDays.get(getAdapterPosition()), binding.chartDay);
        }
    }

    public interface OnDayClickListener {
        void onRowClick(ChartDayModel day, View view);
    }

    @NonNull
    @Override
    public ChartAdapter.ChartViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ChartViewHolder(ChartDayBinding.inflate(LayoutInflater.from(mContext),
                parent, false)
                , mOnDayListener);
    }

    @Override
    public void onBindViewHolder(@NonNull ChartViewHolder holder, int position) {
        ChartDayModel theDay = chartDays.get(position);
        String monthAndYearTogether = mContext.getResources()
                .getStringArray(R.array.Months)[Integer.parseInt(Objects.requireNonNull(theDay.getMonth()))] + "\n" + theDay.getYear();
        holder.binding.textViewDayDayOfMonth.setText(theDay.getDayOfMonth());
        holder.binding.textViewDayMonthAndYear.setText(monthAndYearTogether);
        holder.setTransitionName(theDay.getId());
        final String[] theDayFields = {theDay.getMorningEmpty(), theDay.getMorningFull(),
                theDay.getAfternoonEmpty(), theDay.getAfternoonFull(), theDay.getEveningEmpty(),
                theDay.getEveningFull(), theDay.getNight()};
        final TextView[] holderTextViews = {holder.binding.textViewDayMorningEmpty,
                holder.binding.textViewDayMorningFull, holder.binding.textViewDayAfternoonEmpty,
                holder.binding.textViewDayAfternoonFull, holder.binding.textViewDayEveningEmpty,
                holder.binding.textViewDayEveningFull, holder.binding.textViewDayNight};

        for(int i = 0; i < theDayFields.length; i++) {
            String field = theDayFields[i];
            if (field == null || field.isEmpty() || field.equals("0")) {
                holderTextViews[i].setText("");
                holderTextViews[i].setBackgroundResource(R.drawable.text_frame);
            } else {
                holderTextViews[i].setText(field);
                //Chance background according to value range.
                int warningFlag = GlucoseLevelChecker.checkGlucoseLevel(mContext, i, Integer.parseInt(field));
                if (warningFlag == 2){
                    holderTextViews[i].setBackgroundResource(R.drawable.text_frame_warning_range);
                } else if (warningFlag == 3){
                    holderTextViews[i].setBackgroundResource(R.drawable.text_frame_margin_range);
                } else {
                    holderTextViews[i].setBackgroundResource(R.drawable.text_frame_normal_range);
                }
            }
        }
    }

    @Override
    public int getItemCount() {
        return chartDays.size();
    }
}

File: app/src/main/java/com/vyy/sekerimremake/features/chart/presenter/master/ChartAdapter.kt
package com.vyy.sekerimremake.features.chart.presenter.master

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.vyy.sekerimremake.R
import com.vyy.sekerimremake.databinding.ChartDayBinding
import com.vyy.sekerimremake.features.chart.domain.model.ChartDayModel
import com.vyy.sekerimremake.features.chart.utils.GlucoseLevelChecker
import java.util.*

class ChartAdapter(private val onRowClick: (ChartDayModel, View) -> Unit) :
    ListAdapter<ChartDayModel, ChartAdapter.ChartViewHolder>(ChartComparator()) {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChartViewHolder {
        val binding: ChartDayBinding =
            ChartDayBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ChartViewHolder(binding, onRowClick)
    }

    override fun onBindViewHolder(holder: ChartViewHolder, position: Int) {
        val currentDay = getItem(position)
        holder.bind(currentDay)
    }

    inner class ChartViewHolder(
        private val binding: ChartDayBinding, onRowClick: (ChartDayModel, View) -> Unit
    ) : RecyclerView.ViewHolder(binding.root) {

        init {
            itemView.setOnClickListener {
                onRowClick(getItem(adapterPosition), binding.chartDay)
            }
        }

        fun bind(currentDay: ChartDayModel) {
            val monthAndYearTogether: String =
                itemView.context.resources.getStringArray(R.array.Months)[Objects.requireNonNull<String>(
                    currentDay.month
                ).toInt()] + "\n" + currentDay.year
            binding.apply {
                root.transitionName = currentDay.id
                textViewDayDayOfMonth.text = currentDay.dayOfMonth
                textViewDayMonthAndYear.text = monthAndYearTogether
                val theDayFields = arrayOf(
                    currentDay.morningEmpty,
                    currentDay.morningFull,
                    currentDay.afternoonEmpty,
                    currentDay.afternoonFull,
                    currentDay.eveningEmpty,
                    currentDay.eveningFull,
                    currentDay.night
                )
                val holderTextViews = arrayOf(
                    textViewDayMorningEmpty,
                    textViewDayMorningFull,
                    textViewDayAfternoonEmpty,
                    textViewDayAfternoonFull,
                    textViewDayEveningEmpty,
                    textViewDayEveningFull,
                    textViewDayNight
                )

                //Chance background according to value range.
                var warningFlag: Int
                for (i in theDayFields.indices) {
                    val field = theDayFields[i]
                    if (field == null || field.isEmpty() || field == "0") {
                        holderTextViews[i].text = ""
                        holderTextViews[i].setBackgroundResource(R.drawable.text_frame)
                    } else {
                        holderTextViews[i].text = field
                        warningFlag = GlucoseLevelChecker.checkGlucoseLevel(
                            itemView.context, i, field.toInt()
                        )
                        when (warningFlag) {
                            2 -> {
                                holderTextViews[i].setBackgroundResource(R.drawable.text_frame_warning_range)
                            }
                            3 -> {
                                holderTextViews[i].setBackgroundResource(R.drawable.text_frame_margin_range)
                            }
                            else -> {
                                holderTextViews[i].setBackgroundResource(R.drawable.text_frame_normal_range)
                            }
                        }
                    }
                }
            }
        }
    }

    class ChartComparator : DiffUtil.ItemCallback<ChartDayModel>() {
        override fun areItemsTheSame(oldItem: ChartDayModel, newItem: ChartDayModel): Boolean =
            oldItem === newItem

        override fun areContentsTheSame(oldItem: ChartDayModel, newItem: ChartDayModel): Boolean =
            oldItem.id == newItem.id

    }
}